### PR TITLE
✨ Support api-approved annotation for CRD with k8s group

### LINKED ISSUE #656: Add support for adding annotations to CRD
For kubernetes projects, if we want to use the `sig.k8s.io` group for CRDs, we need to provide the annotation `api-approved.kubernetes.io`. Otherwise, apiserver rejects the CRD.

As a result, we have to manually edit the CRDs produced by `controller-gen` and continuously edit the files every time there is a type update.

Using kustomize is not possible, because the env-test doesn't load them and it could be problematic to do so (it's an additional dependency).

/kind feature

### PR DESCRIPTION
**Pull request by FillZpp**
Support api-approved annotation for CRD with k8s group

fixes #656

### CODE REVIEW

**Comment by FillZpp** [comment]
/retitle ✨ Support api-approved annotation for CRD with k8s group

File: pkg/crd/markers/crd.go [hunk 1]
```diff
@@ -51,6 +52,9 @@ var CRDMarkers = []*definitionWithHelp{
 
 	must(markers.MakeDefinition("kubebuilder:deprecatedversion", markers.DescribesType, DeprecatedVersion{})).
 		WithHelp(DeprecatedVersion{}.Help()),
+
+	must(markers.MakeDefinition("kubebuilder:apiapprovedk8sgroup", markers.DescribesType, APIApprovedKubernetesGroup{})).
```
Comment (alvaroaleman) [inline pkg/crd/markers/crd.go:56]: Why the`k8sgroup` suffix?

Comment (FillZpp) [reply pkg/crd/markers/crd.go:56]: Emm, I supposed it to be short name of the annotation key `api-approved.kubernetes.io`. Do you think `k8sgroupapprovedapi` would be better?

File: pkg/crd/markers/crd.go [hunk 2]
```diff
@@ -345,3 +349,29 @@ func (s DeprecatedVersion) ApplyToCRD(crd *apiext.CustomResourceDefinitionSpec,
...
+// +controllertools:marker:generateHelp:category=CRD
+
+// APIApprovedKubernetesGroup configures the api-approved.kubernetes.io for a CRD
+// with Kubernetes groups, including "k8s.io", "*.k8s.io", "kubernetes.io", "*.kubernetes.io".
+type APIApprovedKubernetesGroup struct {
+	// URL is the value of api-approved.kubernetes.io annotation on CRD,
+	// it should be a valid URL which contains Scheme and Host and
+	// should not start with "unapproved".
+	URL string `marker:"URL"`
+}
```
Comment (joelanford) [inline pkg/crd/markers/crd.go:362]: Any chance we could make this a more generic `Annotation` struct with `Key` and `Value` fields that we parse out of the marker?

I can imagine a few other use cases, for example adding annotation `cert-manager.io/inject-ca-from-secret` to the CRD when it needs a CA injected for a conversion webhook configuration.

Comment (FillZpp) [reply pkg/crd/markers/crd.go:362]: Thanks, it makes sense. I will update this later.

**Comment by FillZpp** [comment]
@joelanford @alvaroaleman I have made it more generic for additional annotations or labels. PTAL :)

File: pkg/crd/testdata/cronjob_types.go [hunk 3]
```diff
@@ -499,6 +499,7 @@ type CronJobStatus struct {
 // +kubebuilder:subresource:status
 // +kubebuilder:resource:singular=mycronjob
 // +kubebuilder:storageversion
+// +kubebuilder:additionalmetadata:annotations="api-approved.kubernetes.io=https://github.com/kubernetes-sigs/controller-tools";"cert-manager.io/inject-ca-from-secret=cert-manager/cert-manager-webhook-ca"
```
Comment (camilamacedo86) [inline pkg/crd/testdata/cronjob_types.go:502]: Why the name is additionalmetada:annotation?

Comment (camilamacedo86) [inline pkg/crd/testdata/cronjob_types.go:502]:
```suggestion
// +kubebuilder:metadata:annotations="api-approved.kubernetes.io=https://github.com/kubernetes-sigs/controller-tools";"cert-manager.io/inject-ca-from-secret=cert-manager/cert-manager-webhook-ca"
```

Do we need the additional? 
Why not only annotations? 
So, if set we add annotations informed 
If not set, we add only the default annotation

Comment (FillZpp) [reply pkg/crd/testdata/cronjob_types.go:502]: Emm.. okey, I will update it to `// +kubebuilder:metadata:annotations=...`

File: pkg/crd/markers/crd.go [hunk 4]
```diff
@@ -51,6 +52,9 @@ var CRDMarkers = []*definitionWithHelp{
 
 	must(markers.MakeDefinition("kubebuilder:deprecatedversion", markers.DescribesType, DeprecatedVersion{})).
 		WithHelp(DeprecatedVersion{}.Help()),
+
+	must(markers.MakeDefinition("kubebuilder:additionalmetadata", markers.DescribesType, AdditionalMetadata{})).
+		WithHelp(AdditionalMetadata{}.Help()),
```
Comment (camilamacedo86) [inline pkg/crd/markers/crd.go:57]: `Are we not adding annotations only?
What about the name here be kubebuilder:annotations

And we Annotations{}

Comment (camilamacedo86) [reply pkg/crd/markers/crd.go:57]: Ps is possible to add annotations outside of the metadata? 
Do we need to use this term here or is that redundant?

Comment (FillZpp) [reply pkg/crd/markers/crd.go:57]: Now it supports to add annotations and labels

```
// +kubebuilder:metadata:annotations=...
// +kubebuilder:metadata:labels=...
```

Not sure if there will be some other fields in metadata have to be added there. So we'd better put all of them in `kubebuilder:metadata`.

File: pkg/crd/testdata/cronjob_types.go (hunk 3 above)
Comment (FillZpp) [reply pkg/crd/testdata/cronjob_types.go]: Updated

**Review by camilamacedo86 — APPROVED** [review]
That seems great for me 🥇 
/lgtm

**Review by alvaroaleman — APPROVED** [review]

**Pull request merged**